From 431305e11d847d93b6d67da8867d272d872c127a Mon Sep 17 00:00:00 2001
From: Jan Engelhardt <jengelh@inai.de>
Date: Wed, 9 Dec 2020 16:58:57 +0100
Subject: [PATCH] Avoid force-encoding display names that fit within qcontent

When the display name contains an At sign, or anything of the sort,
libvmime would forcibly encode this to =?...?=, even if the line
is fine ASCII which only needs quoting.

References: https://github.com/kisli/vmime/pull/246
---
 src/vmime/mailbox.cpp     | 34 ++--------------------------------
 tests/parser/bodyTest.cpp | 13 +++++++++++++
 2 files changed, 15 insertions(+), 32 deletions(-)

diff --git a/src/vmime/mailbox.cpp b/src/vmime/mailbox.cpp
index db5cc74e..30a082ea 100644
--- a/src/vmime/mailbox.cpp
+++ b/src/vmime/mailbox.cpp
@@ -355,44 +355,14 @@ void mailbox::generateImpl(
 		//   - if it contains characters in a charset different from "US-ASCII",
 		//   - and/or if it contains one or more of these special chars:
 		//        CR  LF  TAB  "  ;  ,  <  >  (  )  @  /  ?  .  =  :
-
+		//     these special chars only require quoting, not full encoding
 		// Check whether there are words that are not "US-ASCII"
 		// and/or contain the special chars.
 		bool forceEncode = false;
 
 		for (size_t w = 0 ; !forceEncode && w != m_name.getWordCount() ; ++w) {
 
-			if (m_name.getWordAt(w)->getCharset() == charset(charsets::US_ASCII)) {
-
-				const string& buffer = m_name.getWordAt(w)->getBuffer();
-
-				for (string::const_iterator c = buffer.begin() ;
-				     !forceEncode && c != buffer.end() ; ++c) {
-
-					switch (*c) {
-
-						case '\r':
-						case '\n':
-						case '\t':
-						case ';':
-						case ',':
-						case '<': case '>':
-						case '(': case ')':
-						case '@':
-						case '/':
-						case '?':
-						case '.':
-						case '=':
-						case ':':
-						case '"':
-
-							forceEncode = true;
-							break;
-					}
-				}
-
-			} else {
-
+			if (m_name.getWordAt(w)->getCharset() != charset(charsets::US_ASCII)) {
 				forceEncode = true;
 			}
 		}
diff --git a/tests/parser/bodyTest.cpp b/tests/parser/bodyTest.cpp
index fa396af5..93319212 100644
--- a/tests/parser/bodyTest.cpp
+++ b/tests/parser/bodyTest.cpp
@@ -27,10 +27,23 @@
 VMIME_TEST_SUITE_BEGIN(bodyTest)
 
 	VMIME_TEST_LIST_BEGIN
+		VMIME_TEST(testGenerate_Mailbox)
 		VMIME_TEST(testGenerate_Text)
 		VMIME_TEST(testGenerate_NonText)
 	VMIME_TEST_LIST_END
 
+	void testGenerate_Mailbox() {
+		using namespace vmime;
+
+		// Check that ASCII display names are not encoded more than necessary
+		emailAddress e("a@b.com");
+		auto a = make_shared<mailbox>(text(word("Foo B@r", charsets::US_ASCII)), e);
+		VASSERT_EQ("generate", "\"Foo B@r\" <a@b.com>", a->generate());
+		VASSERT_NEQ("generate", "=?utf-8?Q?Foo_B=40r?= <a@b.com>", a->generate());
+
+		a = make_shared<mailbox>(text(word("Foo B@r", charsets::UTF_8)), e);
+		VASSERT_EQ("generate", "=?utf-8?Q?Foo_B=40r?= <a@b.com>", a->generate());
+	}
 
 	void testGenerate_Text() {
 
